proxy/frontend_handler.go

```diff
@@ -41,28 +42,29 @@ func (h *FrontendHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 		h.closeConnection(ws)
 		return
 	}
+	defer h.backend.closeConnection(hostKey, msgKey)
 
 	// Send response messages to client
 	go func() {
 		for {
 			message, ok := <-respChannel
 			if !ok {
-				h.closeConnection(ws)
 				return
 			}
 			switch message.Type {
 			case common.Body:
-				ws.WriteMessage(1, []byte(message.Body))
+				ws.SetWriteDeadline(time.Now().Add(10 * time.Second))
+				if err := ws.WriteMessage(1, []byte(message.Body)); err != nil {
+					h.closeConnection(ws)
+				}
 			case common.Close:
 				h.closeConnection(ws)
-				return
```

Comment:
This was the critical line causing the bug. This goroutine woiuld exit and then there would be nothing to read from the channel. So now, the only thing that can exit this goroutine is the channel exiting. And the channel is closed via `defer h.backend.closeConnection(hostKey, msgKey)` above